torch/ao/quantization/quantizer/xnnpack_quantizer_utils.py

```diff
@@ -318,129 +325,114 @@ def _annotate_conv_relu(
-    )
-    annotated_partitions = []
-    for fused_partition in fused_partitions:
-        conv_partition, bn_partition, relu_partition = fused_partition
-        annotated_partitions.append(
-            conv_partition.nodes + bn_partition.nodes + relu_partition.nodes
-        )
-        if len(relu_partition.output_nodes) > 1:
-            raise ValueError("Relu partition has more than one output node")
-        relu_node = relu_partition.output_nodes[0]
-        if len(conv_partition.output_nodes) > 1:
-            raise ValueError("conv partition has more than one output node")
-        conv_node = conv_partition.output_nodes[0]
-        conv_node_users = list(conv_node.users.keys())
-        if len(conv_node_users) > 1:
-            raise ValueError(
-                "Conv node must be consumed by BN only for it to be fusable."
-            )
-        if len(bn_partition.output_nodes) > 1:
-            raise ValueError("BatchNorm partition has more than one output node")
-        bn_output_node = bn_partition.output_nodes[0]
+    def _get_pattern(conv_fn: Callable):
+        def _conv_bn_relu(x, conv_weight, conv_bias, bn_weight, bn_bias, bn_rm, bn_rv):
+            conv = conv_fn(x, conv_weight, conv_bias)
+            bn = F.batch_norm(conv, bn_rm, bn_rv, bn_weight, bn_bias, training=True)
+            relu = F.relu(bn)
+            return relu, {"input": x, "conv": conv, "weight": conv_weight, "bias": conv_bias, "output": relu}
+        return _conv_bn_relu
 
-        if _is_annotated([relu_node, bn_output_node, conv_node]):
-            continue
+    return _do_annotate_conv_bn(gm, quantization_config, filter_fn, _get_pattern)
+
+
+def _do_annotate_conv_bn(
+    gm: torch.fx.GraphModule,
+    quantization_config: Optional[QuantizationConfig],
+    filter_fn: Optional[Callable[[Node], bool]],
+    get_pattern: Callable,
+) -> List[List[Node]]:
+    """
+    Given a function that takes in a `conv_fn` and returns a conv-bn[-relu] pattern,
+    return a list of annotated partitions.
+
+    The output of the pattern must include a dictionary from string name to node
+    for the following names: "input", "conv", "weight", "bias", and "output".
+    """
+    # Needed for matching, otherwise the matches gets filtered out due to unused
+    # nodes returned by batch norm
+    gm.graph.eliminate_dead_code()
+    gm.recompile()
+
+    matches = []
+    for conv_fn, example_inputs in [
+        (F.conv2d, _conv2d_bn_example_inputs),
+    ]:
+        pattern = get_aten_graph_module(get_pattern(conv_fn), example_inputs)
+        pattern.graph.eliminate_dead_code()
+        pattern.recompile()
+        matcher = SubgraphMatcherWithNameNodeMap(pattern, ignore_literals=True)
+        matches.extend(matcher.match(gm.graph))
+
+    annotated_partitions = []
+    for match in matches:
+        name_node_map = match.name_node_map
+        input_node = name_node_map["input"]
+        conv_node = name_node_map["conv"]
```

Comment:
nit: to be more robust to decompositions, we'd like to get first user of input_node, first user of weight_node and first user of bias_node etc. and annotate these nodes instead

Reply:
Ok, but isn't that the whole point of `name_node_map`, so that the user doesn't have to manually walk through the graph?

Reply:
the goal is for easier access of nodes, and in this case we want access to input and output of the pattern only, since only these are robust

Reply:
Ok then maybe we should update the tutorials :): https://pytorch.org/tutorials/prototype/pt2e_quantizer.html#a-note-on-ir-for-pt2e-quantization-flow